### PR TITLE
migration-groups: create inventory groups client.

### PR DESCRIPTION
**Pull request by ldjebran**
# Description
 - create inventory groups client
 - add uuid to devices groups models that will correspond to inventory group id, that way we will be aware about the already migrated groups, and choose only those without uuid for migration, also that will create a link between local groups and inventory groups for later hosts adding (as hosts are added by group uuid).

 FIXES: https://issues.redhat.com/browse/THEEDGE-3553

## Type of change

What is it?

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Documentation update
- [ ] Tests update
- [ ] Refactor

# Checklist:

- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes
- [X] I run `make pre-commit` to check fmt/vet/lint/test-no-fdo

### CODE REVIEW

File: pkg/clients/inventorygroups/client.go [hunk 1]
```diff
@@ -0,0 +1,355 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	if requestParams.Page == 0 {
+		requestParams.Page = 1
+	}
+	if requestParams.PerPage == 0 {
+		requestParams.PerPage = DefaultPerPage
+	}
```
Comment (de1987) [inline pkg/clients/inventorygroups/client.go:117]: I would suggest also checking if someone tries to use a negative number.
WDYT?

File: pkg/clients/inventorygroups/client.go [hunk 2]
```diff
@@ -0,0 +1,355 @@
...
+	if res.StatusCode != http.StatusOK {
+		c.log.WithFields(log.Fields{"statusCode": res.StatusCode, "responseBody": string(body)}).Error("inventory groups error response")
+		return nil, ErrGroupsRequestResponse
+	}
+
+	var response Response
+	err = json.Unmarshal(body, &response)
+	if err != nil {
+		c.log.WithField("error", err.Error()).Error("error occurred when unmarshalling response body")
```
Comment (de1987) [inline pkg/clients/inventorygroups/client.go:164]: if we have this type of error, having the body in the logs would be great for investigating the issue.

File: pkg/clients/inventorygroups/client.go [hunk 3]
```diff
@@ -0,0 +1,355 @@
...
+	}
+	response, err := c.ListGroups(ListGroupsParams{Name: name, PerPage: 1, Page: 1, OrderBy: "name", OrderHow: "ASC"})
+	if err != nil {
+		c.log.WithFields(log.Fields{"group-name": name, "error": err.Error()}).Error("failed when calling to ListGroups")
+		return nil, err
+	}
+	if len(response.Results) == 0 || response.Results[0].Name != name {
+		// return group not found when no group returned,
+		// or when the first found group has name different from the requested one.
+		c.log.WithField("group-name", name).Error("group not found")
```
Comment (de1987) [inline pkg/clients/inventorygroups/client.go:184]: If we have a scenario that returns more than one group
Adding the number of responses in the logs would be great for investigating it
WDYT?

File: pkg/clients/inventorygroups/client.go [hunk 4]
```diff
@@ -0,0 +1,355 @@
...
+		c.log.WithFields(log.Fields{"statusCode": res.StatusCode, "responseBody": string(body)}).Error("inventory group error response")
+		return nil, ErrGroupsRequestResponse
+	}
+
+	var response Response
+	err = json.Unmarshal(body, &response)
+	if err != nil {
+		c.log.WithField("error", err.Error()).Error("error occurred when unmarshalling response body")
+		return nil, err
+	}
```
Comment (de1987) [inline pkg/clients/inventorygroups/client.go:233]: if we have this type of error, having the `body` in the logs would be great for investigating the issue.

File: pkg/clients/inventorygroups/client.go [hunk 5]
```diff
@@ -0,0 +1,355 @@
...
+		c.log.WithFields(log.Fields{"statusCode": res.StatusCode, "responseBody": string(body)}).Error("inventory create group error response")
+		return nil, ErrGroupsRequestResponse
+	}
+
+	var group Group
+	err = json.Unmarshal(body, &group)
+	if err != nil {
+		c.log.WithField("error", err.Error()).Error("error occurred when unmarshalling response body")
+		return nil, err
+	}
```
Comment (de1987) [inline pkg/clients/inventorygroups/client.go:292]: if we have this type of error, having the `body` in the logs would be great for investigating the issue.

File: pkg/clients/inventorygroups/client.go [hunk 6]
```diff
@@ -0,0 +1,355 @@
...
+	for key, value := range headers {
+		req.Header.Add(key, value)
+	}
+	client := clients.ConfigureClientWithTLS(&http.Client{})
+	res, err := client.Do(req)
+	if err != nil {
+		c.log.WithField("error", err.Error()).Error("inventory create group request error")
+		return nil, err
+	}
+	defer res.Body.Close()
```
Comment (de1987) [inline pkg/clients/inventorygroups/client.go:275]: I notice this code is repeating in the client implementation, I think we can create a reusable API rest caller function.
WDYT?

File: pkg/models/devicegroups.go [hunk 7]
```diff
@@ -21,6 +21,7 @@ type DeviceGroup struct {
 	Type        string   `json:"Type" gorm:"default:static;<-:create"`
 	Devices     []Device `faker:"-" json:"Devices" gorm:"many2many:device_groups_devices;"`
 	ValidUpdate bool     `json:"ValidUpdate" gorm:"-:all"`
+	UUID        string   `json:"uuid,omitempty" gorm:"index"`
```
Comment (de1987) [inline pkg/models/devicegroups.go:24]: just to confirm, does `UUID` come from inventory?
Thanks

File: pkg/clients/inventorygroups/client.go (hunk 6 above)
Comment (ldjebran) [reply pkg/clients/inventorygroups/client.go:275]: yes you are right , but I do not think that should be implemented in the context of the current PR , also this is the last implementation of a client :smiley:

File: pkg/clients/inventorygroups/client.go (hunk 3 above)
Comment (ldjebran) [reply pkg/clients/inventorygroups/client.go:184]: the page is one and perpage is one too , so the number is always one.

File: pkg/models/devicegroups.go (hunk 7 above)
Comment (ldjebran) [reply pkg/models/devicegroups.go:24]: yes correct, this is the id of the group in inventory /groups and it's in the format of uuid

File: pkg/clients/inventorygroups/client.go (hunk 1 above)
Comment (ldjebran) [reply pkg/clients/inventorygroups/client.go]: done

File: pkg/clients/inventorygroups/client.go (hunk 2 above)
Comment (ldjebran) [reply pkg/clients/inventorygroups/client.go]: done

File: pkg/clients/inventorygroups/client.go (hunk 4 above)
Comment (ldjebran) [reply pkg/clients/inventorygroups/client.go]: done

**Comment by ldjebran** [comment]
@de1987 addressed and responded to all you comment please revisit

File: pkg/clients/inventorygroups/client.go (hunk 3 above)
Comment (ldjebran) [reply pkg/clients/inventorygroups/client.go]: done

**Comment by ldjebran** [comment]
@de1987 thank you for the review, addressed and responded to all you comment please revisit.

**Review by de1987 — APPROVED** [review]

**Review by acosferreira — COMMENTED** [review]
lgtm

**Review by acosferreira — APPROVED** [review]

**Pull request merged**